service/builder.go

```diff
@@ -0,0 +1,267 @@
+						return m(ctx, chain)
+					})
+				}(m)
+			}
+		}
+		if len(descriptor.Definitions) > 0 {
+			for _, d := range descriptor.Definitions {
+				bd.definitions = append(bd.definitions, *b.copyDefinition(&d, consumes, produces))
+			}
+		}
+	}
+	if descriptor.Consumes != nil {
+		consumes = descriptor.Consumes
+	}
+	if descriptor.Produces != nil {
+		produces = descriptor.Produces
+	}
+	for _, child := range descriptor.Children {
+		b.addDescriptor(path, consumes, produces, child)
+	}
+}
+
+func (b *builder) copyDefinition(d *definition.Definition, consumes []string, produces []string) *definition.Definition {
+	// It copy fields except document.
+	newOne := &definition.Definition{}
+	*newOne = *d
+	if d.Consumes != nil {
+		newOne.Consumes = make([]string, len(d.Consumes))
+		copy(newOne.Consumes, d.Consumes)
+	} else if consumes != nil {
+		newOne.Consumes = make([]string, len(consumes))
+		copy(newOne.Consumes, consumes)
+	}
+	if d.Produces != nil {
+		newOne.Produces = make([]string, len(d.Produces))
+		copy(newOne.Produces, d.Produces)
+	} else if produces != nil {
+		newOne.Produces = make([]string, len(produces))
+		copy(newOne.Produces, produces)
+	}
+	newOne.Parameters = make([]definition.Parameter, len(d.Parameters))
+	for i, p := range d.Parameters {
+		p.Description = ""
+		newOne.Parameters[i] = p
+	}
+	newOne.Results = make([]definition.Result, len(d.Results))
+	for i, r := range d.Results {
+		r.Description = ""
+		newOne.Results[i] = r
+	}
+	return newOne
+}
+
+// Build builds a service to handle request.
+func (b *builder) Build() (Service, error) {
+	if len(b.bindings) <= 0 {
+		return nil, noRouter.Error()
+	}
+	var root router.Router
+	for path, bd := range b.bindings {
+		b.logger.V(log.LevelDebug).Infof("Definitions: %d Middlewares: %d Path: %s",
+			len(bd.definitions), len(bd.middlewares), path)
+		router, leaf, err := router.Parse(path)
+		if err != nil {
+			b.logger.Errorf("Can't parse path: %s, %s", path, err.Error())
+			return nil, err
+		}
+		if len(bd.definitions) > 0 {
+			inspector := newInspector(path, b.logger)
+			for _, d := range bd.definitions {
+				b.logger.V(log.LevelDebug).Infof("  Method: %s Consumes: %v Produces: %v",
+					d.Method, d.Consumes, d.Produces)
+				b.modifier(&d)
+				if err := inspector.addDefinition(d); err != nil {
+					return nil, err
+				}
+			}
+			leaf.SetInspector(inspector)
+		}
+		if len(bd.middlewares) > 0 {
+			leaf.AddMiddleware(bd.middlewares...)
+		}
+		if root == nil {
+			root = router
+		} else {
+			root, err = root.Merge(router)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+	s := &service{
+		root:      root,
+		filters:   b.filters,
+		logger:    b.logger,
+		producers: AllProducers(),
+	}
+	for _, p := range producers {
+		s.producers = append(s.producers, p)
+	}
+	s.pool.New = newHTTPContext
+	return s, nil
+}
+
+// Service handles HTTP requests.
+//
+// Workflow:
```

Comment:
Is this still up-to-date?  Is `inspector` part of the workflow?

There is also concepts like `modifier`, etc, which I think is not part of http request workflow (used to build service).  It worths explaining such terminologies.

Reply:
I'll explain it in #2 tonight.